Composer/packages/extensions/obiformeditor/src/Form/fields/CustomObjectArrayField.tsx

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/** @jsx jsx */
+import { jsx } from '@emotion/core';
+import React, { useState } from 'react';
+import formatMessage from 'format-message';
+import { TextField } from 'office-ui-fabric-react/lib/TextField';
+import { IconButton } from 'office-ui-fabric-react/lib/Button';
+import { IContextualMenuItem } from 'office-ui-fabric-react/lib/ContextualMenu';
+import { NeutralColors, FontSizes, SharedColors } from '@uifabric/fluent-theme';
+import { FieldProps } from '@bfcomposer/react-jsonschema-form';
+import { JSONSchema6 } from 'json-schema';
+
+import { swap, remove } from './../utils';
+import { WidgetLabel } from './../widgets/WidgetLabel';
+import { EditableField } from './EditableField';
+import {
+  field,
+  customObjectArrayField,
+  customObjectArrayItem,
+  customObjectArrayItemContainer,
+  customObjectArrayItemLabel,
+  customObjectArrayItemValue,
+  customObjectArrayItemValueLabel,
+} from './styles';
+
+interface ObjectItemProps {
+  index: number;
+  item: { [key: string]: any };
+  itemProperties: JSONSchema6;
+  keyArray: string[];
+  hasMoveUp: boolean;
+  hasMoveDown: boolean;
+  onReorder: (a: number, b: number) => void;
+  onDelete: (idx: number) => void;
+  onEdit: (idx: number, value?: { [key: string]: any }) => void;
+}
+
+const ObjectItem: React.FC<ObjectItemProps> = props => {
+  const { item, index, itemProperties, keyArray = [], onEdit, hasMoveUp, hasMoveDown, onReorder, onDelete } = props;
+  const [key, setKey] = useState(`${item[keyArray[0]]}:${item[keyArray[1]]}`);
+
+  const contextItems: IContextualMenuItem[] = [
```

Comment:
Use `useMemo` to cache. If you see the code below, you will see this `contextItems` just get wrapped up in a `menuProps` prop.

```js
  const menuProps  = useMemo(() => ({ items: [
```